app/javascript/components/student/editor/Settings.tsx

```diff
@@ -0,0 +1,109 @@
+import React, { useState, useRef, useEffect } from 'react'
+import { usePopper } from 'react-popper'
+import { Icon } from '../../common/Icon'
+import { Keybindings } from './types'
+
+export function Settings({
+  theme,
+  keybindings,
+  wrap,
+  setTheme,
+  setKeybindings,
+  setWrap,
+}: {
+  theme: string
+  keybindings: Keybindings
+  wrap: 'off' | 'on' | 'wordWrapColumn' | 'bounded'
+  setTheme: (theme: string) => void
+  setKeybindings: (keybinding: Keybindings) => void
+  setWrap: (wrap: 'off' | 'on' | 'wordWrapColumn' | 'bounded') => void
+}) {
+  const [open, setOpen] = useState(false)
+  const buttonRef = useRef<HTMLButtonElement>(null)
+  const panelRef = useRef<HTMLDivElement>(null)
+  const componentRef = useRef<HTMLDivElement>(null)
+  const { styles, attributes, update } = usePopper(
+    buttonRef.current,
+    panelRef.current,
+    {
+      placement: 'bottom',
+    }
+  )
+
+  const handleClickOutside = (e: MouseEvent) => {
+    if (componentRef.current?.contains(e.target as Node)) {
+      return
+    }
+
+    setOpen(false)
+  }
+
+  useEffect(() => {
+    document.addEventListener('click', handleClickOutside)
+
+    return () => {
+      document.removeEventListener('click', handleClickOutside)
+    }
+  }, [])
```

Comment:
This will work, but it will stop working once you add _enclosed_ values to `handleClickOutside`, because the bound click handler is the one on initial mount, not the one on the current/last render. It's fine for now, but something to look out for. In this case, you can _define_ `handleClickOutside` *inside* the `useEffect` to stop worrying about this:

```typescript
  useEffect(() => {
    const handleClickOutside = (e: MouseEvent) => {
      if (componentRef.current?.contains(e.target as Node)) {
        return
      }

      setOpen(false)
    }
  
    document.addEventListener('click', handleClickOutside)
    return () => {
      document.removeEventListener('click', handleClickOutside)
    }
  }, [componentRef, setOpen])
```

Reply:
Good catch!